carla/recourse_methods/catalog/cem/cem.py

```diff
@@ -0,0 +1,546 @@
+                        CONST[batch_idx] *= 10
+
+        # return the best solution found
+        overall_best_attack = np.array(overall_best_attack[0])
+        overall_best_attack = overall_best_attack.reshape(
+            (1,) + overall_best_attack.shape
+        )
+        return overall_best_attack
+
+    def counterfactual_search(
+        self, instance: np.ndarray
+    ) -> Tuple[np.ndarray, np.ndarray]:
+
+        orig_prob, orig_class, orig_prob_str = self.model_prediction(
+            self.catalog_model.raw_model, np.expand_dims(instance, axis=0)
+        )
+
+        target_label = orig_class
+        orig_sample, target = self.generate_data(instance, target_label)
+
+        # start the search
+        counterfactual = self.attack(orig_sample, target)
+
+        adv_prob, adv_class, adv_prob_str = self.model_prediction(
+            self.catalog_model.raw_model, counterfactual
+        )
+        delta_prob, delta_class, delta_prob_str = self.model_prediction(
+            self.catalog_model.raw_model, orig_sample - counterfactual
+        )
+
+        INFO = "[kappa:{}, Orig class:{}, Adv class:{}, Delta class: {}, Orig prob:{}, Adv prob:{}, Delta prob:{}".format(
+            self.kappa,
+            orig_class,
+            adv_class,
+            delta_class,
+            orig_prob_str,
+            adv_prob_str,
+            delta_prob_str,
+        )
+        print(INFO)
+
+        if np.argmax(
+            self.catalog_model.raw_model.predict(instance.reshape(1, -1))
+        ) != np.argmax(
+            self.catalog_model.raw_model.predict(counterfactual.reshape(1, -1))
+        ):
+            counterfactual = counterfactual
+        else:
+            counterfactual = counterfactual
+            counterfactual[:] = np.nan
+
+        return instance, counterfactual.reshape(-1)
+
+    def get_counterfactuals(self, factuals: pd.DataFrame) -> pd.DataFrame:
+        """
+        Compute a certain number of counterfactuals per factual example.
+
+
+        Parameters
+        ----------
+        factuals : pd.DataFrame
+            DataFrame containing all samples for which we want to generate counterfactual examples.
+            All instances should belong to the same class.
+
+        Returns
+        -------
+
+        """
+        # drop targets
+        target_name = self.data.target
+        instances = factuals.copy()
+
+        continous_cols = self.data.continous
+        fitted_scaler = self.catalog_model.scaler
+        categorical_cols = self.data.categoricals
+        fitted_encoder = self.catalog_model.encoder
+        encoded_features = fitted_encoder.get_feature_names(categorical_cols)
+
+        # normalize and one-hot-encoding
+        instances[continous_cols] = fitted_scaler.transform(instances[continous_cols])
+        instances[encoded_features] = fitted_encoder.transform(
+            instances[categorical_cols]
+        )
+        instances = instances[self.catalog_model.feature_input_order]
```

Comment:
Recourse methods have now a neat inherited property that allows doing this scaling and encoding in one line.

You could call here:

`instances = self.encode_normalize_order_factuals(instances)`

Reply:
thats neat! fixed.